src/language-server/diagnostics.ts

```diff
@@ -33,16 +37,50 @@ function prepareFiles(workspaceFolder: string, documents: TextDocuments<TextDocu
   })
 }
 
-async function compileAndExtractLogs(files: AlgoFile[]): Promise<LogEventWithSource[]> {
+async function compileAndExtractLogs(files: AlgoFile[], puyaService: PuyaService): Promise<LogEventWithSource[]> {
   const logCtx = LoggingContext.create()
+
   await logCtx.run(async () => {
-    await compile(
-      new CompileOptions({
+    try {
+      // Compile with the original compile method but in dry-run mode to get the AWST
+      const compileOptions = new CompileOptions({
         filePaths: files,
-        dryRun: false,
-      }),
-    )
+        dryRun: true,
+      })
+
+      const { awst, compilationSet, programDirectory, ast: sourceFiles } = await compile(compileOptions)
+
+      if (!awst || !compilationSet || !sourceFiles) {
+        return
+      }
+
+      // Serialize AWST and build compilation set mapping for the puya service
+      const serializer = new AwstSerializer({
+        programDirectory: programDirectory,
+        sourcePaths: 'absolute',
+      })
+
+      const serializedAwst = serializer.serialize(awst)
+      const sourceAnnotations = jsonSerializeSourceFiles(sourceFiles, programDirectory)
+
+      const compilationSetMapping = buildCompilationSetMapping({
+        awst,
+        inputPaths: files,
+        compilationSet,
+      })
+
+      // Use the puya service to compile the AWST
+      await puyaService.compile({
+        awst: serializedAwst,
+        compilationSet: compilationSetMapping,
+        sourceAnnotations,
+      })
+    } catch (_) {
```

Comment:
Is the catch even needed now or are we simply keeping just in case compile throws?

Reply:
yeah its mainly in case compile throws exceptions